File: solutions/HJK9810/2026-09-01/high/4014.py
def transit(boards, N):
    for col in range(N):
        for row in range(col):
            boards[row][col], boards[col][row] = boards[col][row], boards[row][col]

    return boards
    

def count_line(line, X):
    is_down = False
    change_count = 1

    for col in range(1, len(line)):
        current = line[col]
        before = line[col - 1]

        if line[col] == line[col - 1]:
            if is_down:
                change_count += 1
                if change_count == X:
                    is_down = False
                    change_count = 0
            else: change_count += 1
        elif abs(current - before) != 1: return False
        elif before < current:
            if is_down or change_count < X: return False
            else: change_count = 1
        elif before > current:
            if is_down: return False

            is_down = X != 1
            change_count = 0 if X == 1 else 1

    return not is_down

def algorithm():
    N, X = map(int, input().split())
    boards = [list(map(int, input().split())) for _ in range(N)]

    count = 0
    for row in range(N):
        if (count_line(boards[row], X)): count += 1

    board_transit = transit(boards, N)
    for row in range(N):
        if (count_line(board_transit[row], X)): count += 1

    return count

T = int(input())
# 여러개의 테스트 케이스가 주어지므로, 각각을 처리합니다.
for test_case in range(1, T + 1):
    print(f"#{test_case} {algorithm()}")

File: solutions/HJK9810/2026-09-01/high/4014.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Solution {
  private boolean checkAirLine(int[] line, int X) {
    boolean isDown = false;
    int changeCount = 1;

    for (int col = 1; col < line.length; col++) {
      int current = line[col];
      int before = line[col - 1];

      if (before == current) {
        if (isDown) {
          changeCount += 1;
          if (changeCount == X) {
            isDown = false;
            changeCount = 0;
          }
        } else {
          changeCount++;
        }
      } else if (Math.abs(current - before) != 1) return false;
      else if (before < current) {
        if (isDown || changeCount < X) return false;
        else changeCount = 1;
      } else if (before > current) {
        if (isDown) return false;

        isDown = X != 1;
        changeCount = X == 1 ? 0 : 1;
      }
    }

    return !isDown;
  }

	public static void main(String args[]) throws Exception {
    Solution sol = new Solution();

		BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;
    int T = Integer.parseInt(input.readLine());

		for(int test_case = 1; test_case <= T; test_case++) {
      st = new StringTokenizer(input.readLine());
      int N = Integer.parseInt(st.nextToken());
      int X = Integer.parseInt(st.nextToken());
      
      int[][] boards = new int[N][N];
      int[][] board_transit = new int[N][N];
      for (int row = 0; row < N; row++) {
        String[] line = input.readLine().split(" ");
        for (int col = 0; col < N; col++) {
          boards[row][col] = Integer.parseInt(line[col]);
          board_transit[col][row] = Integer.parseInt(line[col]);
        }
      }

      int count = 0;
      for (int row = 0; row < N; row++) {
        if (sol.checkAirLine(boards[row], X)) count++;
        if (sol.checkAirLine(board_transit[row], X)) count++;
      }

      System.out.printf("#%d %d\n", test_case, count);
		}
	}
}
